### ISSUE TITLE
Identical completions in multi-node GRPO training

### ISSUE THREAD

**Issue by dmit-vuk** [modelscope/ms-swift#7219]
**Describe the bug**
Hello! I've encountered an issue during multi-node training. In `completions.jsonl`, I'm getting completely identical model responses. I attempted to run your example from [GitHub](https://github.com/modelscope/ms-swift/blob/main/examples/train/grpo/multi_node/server_multi_node.sh): 

I launched a job on 4 nodes and split your file into 4 parts, one for each node.

I start the rollout servers like this:
```bash
CUDA_VISIBLE_DEVICES=0,1,2,3 \
swift rollout \
    --model Qwen/Qwen2.5-VL-3B-Instruct \
    --vllm_tensor_parallel_size 2 \
    --vllm_data_parallel_size 2 \
    --port 8000
```

The training process is done as follows. Master node:
```bash
NNODES=2 \
NODE_RANK=0 \
MASTER_ADDR=127.0.0.1 \
MASTER_PORT=29500 \
CUDA_VISIBLE_DEVICES=0,1,2,3 \
NPROC_PER_NODE=4 \
swift rlhf \
    --rlhf_type grpo \
    --model Qwen/Qwen2.5-VL-3B-Instruct \
    --reward_funcs accuracy \
    --use_vllm true \
    --vllm_mode server \
    --vllm_server_host X.X.X.X X.X.X.X \
    --vllm_server_port 8000 8080 \
    --dataset 'AI-MO/NuminaMath-TIR#1000' \
    --load_from_cache_file true \
    --max_completion_length 2048 \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 1 \
    --learning_rate 1e-6 \
    --save_total_limit 2 \
    --logging_steps 1 \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 4 \
    --dataset_num_proc 4 \
    --num_generations 4 \
    --deepspeed zero2 \
    --log_completions true \
```

And on the second node (worker):
```bash
NNODES=2 \
NODE_RANK=1 \
MASTER_ADDR=X.X.X.X \
MASTER_PORT=29500 \
CUDA_VISIBLE_DEVICES=0,1,2,3 \
NPROC_PER_NODE=4 \
swift rlhf \
    --rlhf_type grpo \
    --model Qwen/Qwen2.5-VL-3B-Instruct \
    --reward_funcs accuracy \
    --use_vllm true \
    --vllm_mode server \
    --vllm_server_host X.X.X.X X.X.X.X \
    --vllm_server_port 8000 8080 \
    --dataset 'AI-MO/NuminaMath-TIR#1000' \
    --load_from_cache_file true \
    --max_completion_length 2048 \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 1 \
    --learning_rate 1e-6 \
    --save_total_limit 2 \
    --logging_steps 1 \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 4 \
    --dataset_num_proc 4 \
    --num_generations 4 \
    --deepspeed zero2 \
    --log_completions true \
```

**Note:** In this case, all responses are unique. However, if I run the rollout server on all 8 GPUs, all responses become repetitive:
```bash
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \
swift rollout \
    --model /mnt/shared_ru.ml.SZ-4_000100/kurtsev/gigavision-ms-swift-train/Qwen2.5-VL-3B-Instruct \
    --vllm_tensor_parallel_size 1 \
    --vllm_data_parallel_size 8 \
    --port 8000
```

I've installed the latest version `ms_swift==3.11.2`, but I'm still getting identical responses during training. Could you please help me understand what I'm doing wrong and how to generalize training for n trainers and m rollouts?

**My hardware and system info**
torch==2.8.0
ms_swift==3.11.2
CUDA Version: 13.0

**Comment by dmit-vuk** [comment]
[Jintao-Huang](https://github.com/Jintao-Huang), help me understand my problem, please

**Comment by dmit-vuk** [comment]
@hjh0119, I tried the latest version 3.12.1 and it doesn’t work either (the answers are completely the same). Please, tell me what I am doing wrong

**Comment by hjh0119** [comment]
I guess this is some kind of seed-related bug (identical random state or random seed across the two nodes).

Maybe you could try modifying the [code](https://github.com/modelscope/ms-swift/blob/v3.12.1/swift/llm/infer/rollout.py#L602-L603) here to ensure the seed differs across devices.

**Comment by dmit-vuk** [comment]
@hjh0119 
If you run [your](https://github.com/modelscope/ms-swift/blob/main/examples/train/grpo/multi_node/server_multi_node.sh) example with the rollouts parameters `CUDA_VISIBLE_DEVICES=0,1,2,3 --vllm_tensor_parallel_size 1 --vllm_data_parallel_size 4`, and the trainer with `--num_generations 8`, then all 8 generated responses will be identical.

The issue is that the config received [here](https://github.com/modelscope/ms-swift/blob/v3.12.1/swift/llm/infer/rollout.py#L602-L603) has no seed value set (`seed == None`). If you add [there](https://github.com/modelscope/ms-swift/blob/a3609c110cee2ec2f26f3d6ccb1cd858dfe9b978/swift/rlhf_trainers/rollout_mixin.py#L115-L125) this parameter `seed=args.seed`, the number of unique generations becomes 4 (each one is repeated twice). This happens because the requests are distributed across different server nodes — 4 requests per node.

I tried setting the parameter `vllm_data_parallel_size=8` in `RLHFArguments`, but it still shows as 4 [here](https://github.com/modelscope/ms-swift/blob/f8421501e7b9fc16780e3aec5954048169d8cb5a/swift/llm/infer/rollout.py#L334). 

Could you please advise how to properly configure the parameters so that all generations are unique?

**Comment by dmit-vuk** [comment]
@hjh0119 
The most likely issue is that the same sample is being sent to different nodes with the rollout server. And with the added line when creating the request_config (`seed=args.seed`), identical requests with the same seed parameter value are sent to different nodes. Please tell me, how can I bind a specific request to a single node of the rollout server? Or is there a way to send a request with a seed that depends on the node's rank?

I tried running rollouts with different seed values (13 and 137):

```
CUDA_VISIBLE_DEVICES=0,1,2,3 \
swift rollout \
    --model Qwen/Qwen2.5-VL-3B-Instruct \
    --vllm_tensor_parallel_size 1 \
    --vllm_data_parallel_size 4 \
    --port 8000 \
    --seed 13
```

But all the generations were still identical.

**Comment by dmit-vuk** [comment]
@hjh0119 
Unfortunately, I still couldn't find the right parameters to make all generations unique.

However, I managed to slightly modify your code to make everything work correctly:
- First, you need to add the line `seed=args.seed` [here](https://github.com/modelscope/ms-swift/blob/a3609c110cee2ec2f26f3d6ccb1cd858dfe9b978/swift/rlhf_trainers/rollout_mixin.py#L115-L125) to pass the seed to `request_config`.
- Second, you need to specify the node rank in the rollout launch configuration. Example:

```bash
CUDA_VISIBLE_DEVICES=0,1,2,3 \
RANK=0 \
swift rollout \
    --model Qwen/Qwen2.5-VL-3B-Instruct \
    --vllm_tensor_parallel_size 1 \
    --vllm_data_parallel_size 4 \
    --port 8000
```

- Third, you need to remember it [here](https://github.com/modelscope/ms-swift/blob/976d82faa1456ea6e07090fc6f9f1dc57f4bc415/swift/pipelines/base.py#L17-L24) (add the line `self.rank = getattr(args, 'rank', -1)`)
- And finally, you need to increment the seed value in the rollout on different nodes based on the rank [here](https://github.com/modelscope/ms-swift/blob/v3.12.1/swift/llm/infer/rollout.py#L593) (`request_config.seed += self.rank * len(infer_requests)`)

**BUT this solution has several limitations!!!**
- In theory, it works if the sampler or vLLM (or whatever is responsible for distributing generations between nodes) divides them equally across each rollout.
- Accordingly, this only works if the number of generations (`num_generations`) is exactly divisible by the number of nodes running the rollout server (`len(vllm_server_host)`).

Could you please suggest a more correct way to do this? Or is it still possible to set some parameters correctly without changing the code?

**Comment by hjh0119** [comment]
maybe fixed in https://github.com/modelscope/ms-swift/pull/7858

**Comment by dmit-vuk** [comment]
@hjh0119 
Yes, really, now it seems to work and all the completions are unique. Thank you very much!
